include/boost/geometry/algorithms/detail/envelope/box.hpp

```diff
@@ -40,36 +42,124 @@ namespace detail { namespace envelope
 {
 
 
-struct envelope_box_on_spheroid
+template
+<
+    std::size_t Index,
+    std::size_t Dimension,
+    std::size_t DimensionCount,
+    typename CS_Tag
+>
+struct envelope_indexed_box
 {
-    template<typename BoxIn, typename BoxOut>
+    template <typename BoxIn, typename BoxOut>
     static inline void apply(BoxIn const& box_in, BoxOut& mbr)
     {
-        BoxIn box_in_normalized = detail::return_normalized<BoxIn>(box_in);
+        detail::indexed_point_view<BoxIn const, Index> box_in_corner(box_in);
+        detail::indexed_point_view<BoxOut, Index> mbr_corner(mbr);
+
+        detail::conversion::point_to_point
+            <
+                detail::indexed_point_view<BoxIn const, Index>,
+                detail::indexed_point_view<BoxOut, Index>,
+                Dimension,
+                DimensionCount
+            >::apply(box_in_corner, mbr_corner);
+    }
+};
 
-        geometry::transform(box_in_normalized, mbr);
+template
+<
+    std::size_t Index,
+    std::size_t DimensionCount
+>
+struct envelope_indexed_box<Index, 0, DimensionCount, spherical_equatorial_tag>
+{
+    template <typename BoxIn, typename BoxOut>
+    static inline void apply(BoxIn const& box_in, BoxOut& mbr)
+    {
+
+       // transform() does not work with boxes of dimension higher
+        // than 2; to account for such boxes we transform the min/max
+        // points of the boxes using the indexed_point_view
+        detail::indexed_point_view<BoxIn const, Index> box_in_corner(box_in);
+        detail::indexed_point_view<BoxOut, Index> mbr_corner(mbr);
+
+        // first transform the units
+        transform_units(box_in_corner, mbr_corner);
```

Comment:
Exactly, in spherical and geographic CS the `Dimension` must be `0` otherwise the template will not work as intended. I understand that this follows the overall design, and that this template is consistent with the rest. This may be an indicator that something is wrong with the whole design.

I don't know how do the others see it but I see it like this. Let's consider a function:

    // \param p1   it may be any number but if p2 is 2 or 3 then p1 must be 0
    // \param p2   it must be 1, 2 or 3
    do_something(p1, p2);

Do you like the design of this function?
The problem is that the parameters are not orthogonal. It's more general actually, it's about a well-defined purpose, defined by the parameters (see e.g. why is `realloc()` considered as a badly designed function). What is true for functions is also true for meta-functions. In the current design the specializations has 2 purposes instead of one. They in the same time pick up the correct algorithm for the CS and implements this algorithm. Actually this would be a better example:

    // \param first  the iterator pointing to the first element of a Range
    //               if the pred is Abc or Def then first must be the "true"
    //               first element of a Range.
    // \note    Have in mind that you should always pass the whole Range here.
    for_each_something(first, last, pred)

It's even more complicated because it's recursive. Maybe a recursive function traversing some tree but for some parameters requireing that you start at the root and for other parameters don't requireing it but then ignoring this other parameter would be a better example.

And I'm not saying that you should support all of the special cases. It doens't have sense because always the same parameters are passed into these templates. I'm saying that for me the design is overcomplicated.